File: DSAlgorithms/Leetcode/c++/arrays_hash/309-best_time_buy_sell_stock_cooldown.cpp
class Solution {
public:
        int dfs(vector<int>& prices, int index, int buying, std::vector<std::vector<int>>& res)
        {
            if(index >= prices.size() )
                return 0;    
            if(res[index][buying] != -1)
                return res[index][buying];
            int cooldown = dfs(prices, index + 1, buying, res);
            if(buying == 1)
            {
                int buy = dfs(prices, index + 1, 0, res) - prices[index];
                res[index][buying] = std::max(buy, cooldown);
            }
            else if(buying == 0)
            {
                int sell = dfs(prices, index + 2, 1, res) + prices[index];
                res[index][buying] = std::max(sell, cooldown);
            }
            return res[index][buying];
        }
    int maxProfit(vector<int>& prices) {
        int prices_size = prices.size();
        std::vector<std::vector<int>> results_dp(prices_size, vector<int>(2, -1));
        /* Does the same as above, init a 2d with prices_size rows and 
        2 cols filled with -1s
        std::vector<std::vector<int>> results_dp(prices_size);
        for(int row = 0; row < prices_size; row++)
        {
            results_dp[row] = std::vector<int> (2);
            for(int col = 0; col < results_dp[row].size(); col++)
                results_dp[row][col] = -1;
        }*/

        return dfs(prices, 0, 1, results_dp);
    }
};

File: dsa_algo/leetcode/ts/src/arrays_hash/309-best_time_buy_sell_stock_cooldown.ts
const dfs = (prices: number[], index: number, buying: number, res: number[][]): number => {
    if(index >= prices.length)
        return 0;
    if(res[index][buying] !== -1)
        return res[index][buying];
    const cooldown = dfs(prices, index + 1, buying, res);
    if(buying === 1)
    {
        const buy = dfs(prices, index + 1, 0, res) - prices[index];
        res[index][buying] = Math.max(buy, cooldown);
    }
     else if(buying === 0)
    {
        const sell = dfs(prices, index + 2, 1, res) + prices[index];
        res[index][buying] = Math.max(sell, cooldown);
    }
    return res[index][buying];
}
const maxProfit = (prices: number[]): number => {
    const prices_size: number = prices.length;
    const results_dp: number[][] = new Array(prices.length);
    for(let i = 0; i < prices_size; i++)
        results_dp[i] = new Array(2).fill(-1);

    return dfs(prices, 0, 1, results_dp);

};